### PR TITLE
[RFC] Improve tests

### PR DESCRIPTION
**Pull request by Shougo**
https://github.com/neovim/neovim/pull/14534

I have fixed the tests in the first.

### CODE REVIEW

**Comment by Shougo** [comment]
I think it is not related my changes.

```
stack traceback:
	test\helpers.lua:73: in function 'eq'
	test/functional\core\job_spec.lua:359: in function <test/functional\core\job_spec.lua:353>


 53 SKIPPED TESTS
 1 ERROR
-- Tests exited non-zero: 1
-- Output to stderr:
```

**Comment by Shougo** [comment]
@janlazo Can you check it?

**Comment by jamessan** [comment]
What's the reasoning behind these test changes?  The tests are running cleanly already, so why do we need to make changes that diverge from Vim?

**Comment by Shougo** [comment]
> What's the reasoning behind these test changes? The tests are running cleanly already, so why do we need to make changes that diverge from Vim?

OK.  I have added some `cd` to test it locally.
If cwd is not home, it seems failed.  I think the tests should be cwd independent.

And I have added some `delete()` to test it locally.
If you execute it locally, previous test files my exists.
I think `delete()` is better.

**Comment by Shougo** [comment]
So current tests are tested/created for only CI or cleaned environment.
I think tests can be executed locally to tests patches.

Am I mistake??

**Comment by janlazo** [comment]
This means that there are unmerged vim patches to delete old test data.

**Comment by Shougo** [comment]
> This means that there are unmerged vim patches to delete old test data.

No.  You can check Vim's source code.
My test fixes are not patch porting.

File: src/nvim/testdir/test_registers.vim [hunk 1]
```diff
@@ -43,6 +43,9 @@ func Test_yank_shows_register()
 endfunc
 
 func Test_display_registers()
+    " Disable clipboard
+    let g:clipboard = {}
```
Comment (janlazo) [inline src/nvim/testdir/test_registers.vim:47]: This is fine but put in separate commit.

Comment (Shougo) [reply src/nvim/testdir/test_registers.vim:47]: Done.

**Comment by Shougo** [comment]
@janlazo Ping.  Can I port the patch?

**Comment by janlazo** [comment]
Yes.

File: src/nvim/testdir/test_quickfix.vim [hunk 2]
```diff
@@ -1012,9 +1012,9 @@ endfunc
 func Test_efm_dirstack()
   " Create the directory stack and files
   call mkdir('dir1')
-  call mkdir('dir1/a')
-  call mkdir('dir1/a/b')
-  call mkdir('dir1/c')
+  call mkdir('dir1/a', 'p')
+  call mkdir('dir1/a/b', 'p')
+  call mkdir('dir1/c', 'p')
```
Comment (janlazo) [inline src/nvim/testdir/test_quickfix.vim:1017]: Not seeing the value of the `'p'` flag when the previous `mkdir()` aren't collapsed to 1 line.
This isn't in Vim but it can be upstreamed.

File: src/nvim/testdir/test_command_count.vim [hunk 3]
```diff
@@ -103,6 +103,8 @@ endfunc
 
 func Test_command_count_2()
   silent! %argd
+  cd
```
Comment (janlazo) [inline src/nvim/testdir/test_command_count.vim:106]: Vim doesn't have this.
`:cd` changes to the home directory in Unix system.
Why?

File: src/nvim/testdir/test_arglist.vim [hunk 4]
```diff
@@ -1,5 +1,10 @@
 " Test argument list commands
 
+func Reset_arglist()
+  cd
```
Comment (janlazo) [inline src/nvim/testdir/test_arglist.vim:4]: https://github.com/neovim/neovim/pull/14540#discussion_r636577072
Same comment about `cd` for Unix systems.

File: src/nvim/testdir/test_registers.vim (hunk 1 above)
Comment (janlazo) [reply src/nvim/testdir/test_registers.vim:47]: Forgot to ask.
Does this affect the other tests in this file in case that this fails?

**Comment by janlazo** [comment]
@Shougo It's beneficial for both projects for you to upstream your improvements to the tests. Don't use `:cd` without args because the documented behavior suggests that it's not portable on non-Unix systems (unless the doc is wrong now).

File: src/nvim/testdir/test_command_count.vim (hunk 3 above)
Comment (Shougo) [reply src/nvim/testdir/test_command_count.vim:106]: I will remove it.

File: src/nvim/testdir/test_registers.vim (hunk 1 above)
Comment (Shougo) [reply src/nvim/testdir/test_registers.vim:47]: Please see it.

https://github.com/neovim/neovim/pull/12784#issuecomment-683495832

It is neovim only problem.

**Comment by Shougo** [comment]
Hm...


`let $GZIP = ""` does not work for me.
`[,']!GZIP= gzip -d` works instead.

**Comment by Shougo** [comment]
Almost tests are passed.

File: test/functional/legacy/011_autocommands_spec.lua [hunk 5]
```diff
@@ -67,10 +67,15 @@ describe('file reading, writing and bufnew and filter autocommands', function()
 
     it('FileReadPost (using gzip)', function()
       prepare_gz_file('Xtestfile', text1)
-      feed_command('let $GZIP = ""')
       --execute('au FileChangedShell * echo "caught FileChangedShell"')
       feed_command('set bin')
-      feed_command("au FileReadPost    *.gz   '[,']!gzip -d")
+      if iswin() then
+          feed_command('let $GZIP = ""')
```
Comment (janlazo) [inline test/functional/legacy/011_autocommands_spec.lua:73]: `call setenv('GZIP', v:null)` for both.

Comment (Shougo) [reply test/functional/legacy/011_autocommands_spec.lua:73]: Unfortunately, it does not work for me.
So I cannot accept the suggestion.

```
stack traceback:
	test/helpers.lua:73: in function 'expect'
	test/functional/legacy/011_autocommands_spec.lua:76: in function <test/functional/legacy/011_autocommands_spec.lua:68>


 1 ERROR
-- Tests exited non-zero: 1
-- Output to stderr:
gzip: warning: GZIP environment variable is deprecated; use an alias or script
Xtestfile:	 73.9% -- replaced with Xtestfile.gz
gzip: warning: GZIP environment variable is deprecated; use an alias or script
Xtestfile:	 73.9% -- replaced with Xtestfile.gz
gzip: warning: GZIP environment variable is deprecated; use an alias or script
Xtestfile:	 73.9% -- replaced with Xtestfile.gz
```

Comment (janlazo) [reply test/functional/legacy/011_autocommands_spec.lua:73]: Does your shell set `GZIP`? Does the problem remain if you change the shell to `sh`?

Comment (Shougo) [reply test/functional/legacy/011_autocommands_spec.lua:73]: OH, it fixes my problem!
Thanks.

**Comment by Shougo** [comment]
I have pushed the fix.

File: src/nvim/testdir/test_registers.vim (hunk 1 above)
Comment (janlazo) [reply src/nvim/testdir/test_registers.vim:47]: Without your changes, I can reproduce that error on my machine depending on the clipboard state.

What I'm asking is if it impacts the other tests in case `Test_display_registers()` because `g:clipboard` is not restored. Vim tests typically restore the previous state of the test runner within the test such that the test must pass to not break the next tests.
It's fine if you don't know.

Comment (Shougo) [reply src/nvim/testdir/test_registers.vim:48]: OK.  I have added the restore.

**Comment by Shougo** [comment]
Can you review it again?

**Comment by Shougo** [comment]
Almost succeeded.

**Comment by Shougo** [comment]
It should be only change tests instead.

**Comment by Shougo** [comment]
I have rebased tests.  Please review it again.

**Comment by Shougo** [comment]
Some tests are failed in Windows CI.
But they are same with master.  My changes are not affected.

**Comment by Shougo** [comment]
Can you review it and merge?

File: src/nvim/testdir/test_arglist.vim [hunk 6]
```diff
@@ -26,6 +30,8 @@ func Test_argidx()
 endfunc
 
 func Test_argadd()
+  call Reset_arglist()
```
Comment (janlazo) [inline src/nvim/testdir/test_arglist.vim:33]: I don't see the rationale in the commit message and it's not in Vim.
What's the issue?

File: src/nvim/testdir/test_arglist.vim [hunk 7]
```diff
@@ -115,8 +121,7 @@ endfunc
 " Test for [count]argument and [count]argdelete commands
 " Ported from the test_argument_count.in test script
 func Test_argument()
-  " Clean the argument list
-  arga a | %argd
+  call Reset_arglist()
```
Comment (janlazo) [inline src/nvim/testdir/test_arglist.vim:124]: Can you send this to Vim?

**Comment by janlazo** [comment]
My only issue is the `Reset_arglist()`  changes that Vim does not have in some tests. I'm fine with the rest.

**Comment by Shougo** [comment]
`Reset_arglist()` seems not needed.

File: src/nvim/testdir/test_arglist.vim (hunk 6 above)
Comment (janlazo) [reply src/nvim/testdir/test_arglist.vim:33]: I suppose that this is to match `Test_zero_argadd`.

Comment (Shougo) [reply src/nvim/testdir/test_arglist.vim:33]:
```
  From test_command_count.vim:
  Found errors in Test_command_count_2():
  function RunTheTest[37]..Test_command_count_2 line 9: Expected 'a' but got ''
```

I think arglist reset is needed.

**Comment by Shougo** [comment]
Done.

https://github.com/vim/vim/pull/9154

File: src/nvim/testdir/test_arglist.vim (hunk 7 above)
Comment (Shougo) [reply src/nvim/testdir/test_arglist.vim:124]: Done.

**Comment by Shougo** [comment]
Merged in https://github.com/vim/vim/commit/3cad47038547e694cfa26ba39c399f610d2054bd

**Comment by Shougo** [comment]
@janlazo I think it has no problem.

**Comment by janlazo** [comment]
Redo one of the commits as a vim-patch and I'll merge.

**Comment by Shougo** [comment]
I have added vim-patch 8.2.3616 commit.

**Comment by janlazo** [comment]
Thank you for your patience and cooperation.

**Comment by Shougo** [comment]
Thank you!
